drivers/can/can_native_posix_linux.c

```diff
@@ -0,0 +1,436 @@
+/*
+ * Copyright (c) 2022 Martin Jäger <martin@libre.solar>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_native_posix_linux_can
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/drivers/can.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/net_pkt.h>
+#include <zephyr/net/socketcan.h>
+#include <zephyr/net/socketcan_utils.h>
+
+#include "can_utils.h"
+#include "can_native_posix_linux_socketcan.h"
+
+LOG_MODULE_REGISTER(can_npl, CONFIG_CAN_LOG_LEVEL);
+
+struct can_filter_context {
+	can_rx_callback_t rx_cb;
+	void *cb_arg;
+	struct can_filter filter;
+};
+
+struct can_npl_data {
+	struct can_filter_context filters[CONFIG_CAN_MAX_FILTER];
+	struct k_mutex filter_mutex;
+	struct k_sem tx_idle;
+	struct k_sem tx_done;
+	can_tx_callback_t tx_callback;
+	void *tx_user_data;
+	bool loopback;
+	bool mode_fd;
+	int dev_fd; /* Linux socket file descriptor */
+	struct k_thread rx_thread;
+
+	K_KERNEL_STACK_MEMBER(rx_thread_stack, CONFIG_ARCH_POSIX_RECOMMENDED_STACK_SIZE);
+};
+
+struct can_npl_config {
+	const char *if_name;
+};
+
+static void dispatch_frame(const struct device *dev, struct can_frame *frame)
+{
+	struct can_npl_data *data = dev->data;
+	can_rx_callback_t callback;
+	struct can_frame tmp_frame;
+
+	k_mutex_lock(&data->filter_mutex, K_FOREVER);
+
+	for (int filter_id = 0; filter_id < ARRAY_SIZE(data->filters); filter_id++) {
+		if (data->filters[filter_id].rx_cb == NULL) {
+			continue;
+		}
+
+		if (!can_utils_filter_match(frame,
+					    &data->filters[filter_id].filter)) {
+			continue;
+		}
+
+		/* Make a temporary copy in case the user modifies the message */
+		tmp_frame = *frame;
+
+		callback = data->filters[filter_id].rx_cb;
+		callback(dev, &tmp_frame, data->filters[filter_id].cb_arg);
+	}
+
+	k_mutex_unlock(&data->filter_mutex);
+}
+
+static void rx_thread(void *arg1, void *arg2, void *arg3)
+{
+	const struct device *dev = arg1;
+	struct can_npl_data *data = dev->data;
+	struct socketcan_frame sframe;
+	struct can_frame frame;
+	bool msg_confirm;
+	int count;
+
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	LOG_DBG("Starting Linux SocketCAN RX thread");
+
+	while (true) {
+		while (linux_socketcan_poll_data(data->dev_fd) == 0) {
+			count = linux_socketcan_read_data(data->dev_fd, (void *)(&sframe),
+							   sizeof(sframe), &msg_confirm);
+			if (msg_confirm) {
+				if (data->tx_callback != NULL) {
+					data->tx_callback(dev, 0, data->tx_user_data);
+				} else {
+					k_sem_give(&data->tx_done);
```

Comment:
In case the `msg_confirm` is not received, `tx_done` semaphore would not be released and the `tx_callback` would not be called. Also, `tx_idle` would not be released.
Would it make sense to use the 'tx_done' semaphore with a  short timeout to be able to report this as transmission error via the callback or as return value in case no callback was passed?
I am not sure how long it takes until the confirmation is expected to be received.

Reply:
But isn't that what the API docs describe? Block until the message is sent. If the message is never sent to the bus, `can_send` will block for ever.

How would you suggest to implement that timeout function? With a kernel timer that gives the tx_done semaphore? And what would be a "short timeout" for you?

Reply:
I found [this thread](https://www.spinics.net/lists/linux-can/msg07758.html) about the reliability of the `MSG_CONFIRM`. It looks like there are ways to determine if messages were dropped under heavy load, but it looks rather complicated with the Socket API.

As this driver is mainly meant for development and testing purposes I think it's acceptable to not cover all edge cases, so I suggest to keep it as is.

Reply:
> But isn't that what the API docs describe? Block until the message is sent. If the message is never sent to the bus, `can_send` will block for ever.

Correct. That's the expected behaviour when calling `can_send()` without a callback.

Reply:
> As this driver is mainly meant for development and testing purposes I think it's acceptable to not cover all edge cases, so I suggest to keep it as is.

I agree.

Reply:
> But isn't that what the API docs describe?

You are right. I just thought we should avoid blocking forever, but introducing an arbitrary timeout is also not great at all.
Thanks for that work! LGTM in this release.